File: ng2-dart/lib/auth.guard.dart
import 'package:angular2/core.dart';
import 'package:angular2/router.dart';

import 'dart:html';

import 'model/user.dart';

import 'service/auth.service.dart';

@Injectable()
class AuthGuard {
	
	final AuthService _authService;
	final Router _router;

	AuthGuard(this._authService, this._router);

	canActivate() {
		var creds = window.localStorage['creds'];

		if (creds != null) {

			var user = new User();
			var credString = window.atob(creds);
			var split = credString.split(":");

			user.username = split[0];
			user.password = split[1];

			this._authService.signIn(user);

			if (_authService.signedIn) {
				return true;
			} else {
				return false;
			}
		} else {
			window.localStorage.remove('creds');
			_router.navigate(['sign-in']);
			return false;
		}
	}
}

canActivateFn(a, b) {
	Injector ij = new Injector();
	AuthGuard ag = ij.get(AuthGuard);

	return ag.canActivate();
}

File: ng2-ts/src/app/shared/auth.guard.ts
import { Injectable } from '@angular/core';
import { Router, CanActivate, ActivatedRouteSnapshot, RouterStateSnapshot} from '@angular/router';
import { AuthService } from '../service/auth.service';

import { User } from '../model/user';

@Injectable()
export class AuthGuard implements CanActivate {
  constructor(private router: Router, private authService: AuthService) {}

  canActivate(route: ActivatedRouteSnapshot, state: RouterStateSnapshot) {  	

    let creds = localStorage.getItem('creds');

    if (creds != null) {
    	if (this.authService.isSignedIn()) {
    		return true;
    	} else {

    		let url = state.url;
    		let user = new User();
       		let credString = atob(creds);
       		let split = credString.split(":");

       		user.username = split[0];
       		user.password = split[1];

      		this.authService.signIn(user)
           		.subscribe((User:any) => {
           			this.router.navigate([url]);
				}, (error:any) => {
					this.router.navigate(['sign-in']);
				});

			return false;
    	}
    } else {
    	localStorage.removeItem('creds');
    	if (this.router.url !== '/sign-in') {
    		this.router.navigate(['sign-in']);
    		return false;
    	} else {
    		return true;
    	}
    }
  }
}
